pkg/datapath/linux/routing/migrate.go

```diff
@@ -0,0 +1,623 @@
+	//
+	// If a failure occurs at (1), then the old state can continue to service
+	// the endpoint. Similarly with (2) because routes without rules are likely
+	// to not have any effect.
+	//
+	// If a failure occurs at (3), we have already succeeded in getting the new
+	// state in-place to direct traffic for the endpoint. In any case of
+	// upgrade or downgrade, it is possible for both states to be in-place if
+	// there are any failures, including any failures in reverting. The
+	// datapath selected will depend on the rule priority. Concretely, for
+	// upgrades, the newer rule will have a lower priority, so the original
+	// datapath will be selected. For downgrades, the newer rule will have a
+	// higher priority, so the newer datapath will be selected. In either case,
+	// no connectivity is affected for the endpoint.
+	//
+	// If we fail at (4), then the old rule will have been deleted and the new
+	// state is in-place, which would be servicing the endpoint. The old routes
+	// would just be leftover state to be cleaned up at a later point.
+	//
+	// It is also important to note that we only revert what we've done on a
+	// per-rule basis if we fail at (2) or (3). This is by design because we
+	// want to ensure that each iteration of the loop is atomic to each
+	// endpoint. Meaning, either the endpoint ends up with the new datapath or
+	// the original.
+
+	var (
+		// Number of rules (endpoints) successfully migrated and how many failed.
+		migrated, failed int
+		// Store the routes to cleanup in a set after successful migration
+		// because routes are only unique per table ID, meaning many endpoints
+		// share the same route table if the endpoint's IP is allocated from
+		// the same ENI device.
+		cleanup = make(map[netlink.Rule][]netlink.Route)
+		// Store the table IDs of the routes whose migration failed. This is
+		// important because to prevent deleting routes for endpoints that
+		// share the same table ID. An example: let's say we have 2 endpoints
+		// that have rules and routes that refer to the same table ID. If 1
+		// endpoint fails the migration and the other succeeded, we must not
+		// remove the routes for the endpoint that failed because it's still
+		// relying on them for connectivity.
+		failedTableIDs = make(map[int]struct{})
+	)
+
+	if isUpgrade {
+		for _, r := range v1Rules {
+			// Let's say we have an ENI device attached to the node with
+			// ifindex 3 and interface number 2. The following rule will exist
+			// on the node _before_ migration.
+			//   110:    from 192.168.11.171 to 192.168.0.0/16 lookup 3
+			// After the migration, this rule will become:
+			//   111:    from 192.168.11.171 to 192.168.0.0/16 lookup 12
+			// The priority has been updated to 111 and the table ID is 12
+			// because the interface number is 2 plus the routing table offset
+			// (linux_defaults.RouteTableInterfacesOffset). See copyRoutes()
+			// for what happens with routes.
```

Comment:
nit, these comments are probably more useful in the `upgradeRule()` logic (and corresponding `downgradeRule()` logic) now that there's a dedicated function. But they're super useful to compare against to reason about the logic :+1:

Reply:
Good point, fixed.